### ISSUE TITLE
ERROR: could not append to file: No space left on device   while running insert query using repartition join

### ISSUE THREAD

**Issue by kot8891** [citusdata/citus#7249]
```
[2023-10-06 10:40:13] [53100] ERROR: could not append to file: No space left on device
[2023-10-06 10:40:13] Where: while executing command on dbwork4-1:5432

```
Configuration 

Postgresql:
PostgreSQL 16.0 (Debian 16.0-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit

CITUS:
Citus 12.1.0 on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
citus.enable_repartition_joins=on

Hardware:
Coordinator node: 6 CPU, 8GB RAM, 20 GB available storage space
4 worker nodes each: 4 CPU, 16 GB RAM 300GB available storage space

The problem is:
while running the query bellow I have dramatically decrease free disk space on all my worker nodes.

```
insert into my_flat_table
select
		dhc.id
	,	left(dscpctikcvced.text_val1, 4) || '**** ****' || right(dscpctikcvced.text_val3, 4)
	,	case
			when dscpctikcvced.text_val4 = 'T'
				then	'ssssss'
			else	'rrrrrr'
		end										                                                    as _type
	,	dscpctikcvced.date_val2
	,	dscpctikcvced.text_val5
	,	coalesce(dscpctikcvced.text_val5, 'EEEEEEE')
	,	dscpctikcvced.date_val1
	,	dscpctikcvced.date_val2
	,	dscssi.text_val1
	,	dscssi.text_val3
from
		table1_dist																	dhc
		join	table2_dist																dscpctikcvced
		on	dscpctikcvced.m_hashkey = dhc.m_hashkey
			join	table3_dist															dscssi
			on	dscssi.m_hashkey = dhc.m_hashkey
				join	table4_dist														dlc2cp
				on		dlc2cp.m_hashkey2 = dhc.m_hashkey
```

I have 4 joins. table1_dist, table2_dist and table3_dist are columnar and collocated and table4_dist is heap table and is not collocated with others. Every table has not more than 3 million rows and total used space by them is not more than 500 MB.
But after start query every second on every worker I see decreasing available space by around 1 GB per second. 
 
![image](https://github.com/citusdata/citus/assets/81918162/a62336f4-537b-42d9-935e-a7d4f91f4941)

Some minutes later I have 
```
[53100] ERROR: could not append to file: No space left on device
```
How to reproduce: 
All simple scripts to reproduce the error are in the attached file.

[sql_to_reproduce.txt](https://github.com/citusdata/citus/files/12840761/sql_to_reproduce.txt)

**Comment by onurctirtir** [comment]
The query plan doesn't change based on "citus.enable_repartition_joins" because the query doesn't have a join that requires doing a "repartitioned join", but it requires doing a "repartitioned insert .. select", which can be controlled by "citus.enable_repartitioned_insert_select". It requires "repartitioned insert .. select" because the query inserts "id" column of "table1_dist" into "id" column of "my_flat_table" and while "id" is the distribution column of "my_flat_table", this is not the case for "table1_dist".

You can disable "citus.enable_repartitioned_insert_select" but this would mean pulling result of the SELECT query to the coordinator and then sending the returned tuples back to the workers; though you might then experience excessive memory / disk usage problem on the coordinator. Even more, "repartitioned insert .. select" is faster than "pull to coordinator" method.

You might want to shard your other tables based on "id" column if that fits to your data model. I don't see any better option other than this.

Feel free to re-open the issue if those don't help to you.

**Comment by onurctirtir** [comment]
First of all, the query cannot change by enabling /disabling repartition joins because the SELECT part of the query doesn't have a join that requires repartition join. But the whole insert .. select query requires "repartitioned insert .. select" because the query doesn't reference the distribution column of one of the source tables while inserting into distribution column of the target table.

So I'm closing the other issue while investigating this one further. (https://github.com/citusdata/citus/issues/7250)

**Comment by onurctirtir** [comment]
In the original query plan, Citus executes 32 different queries in parallel for the SELECT part of the query and I converted it into a COUNT(1) query to simplify it. Then picked one of those 32 queries executed on shards.

And I've seen that even a single shard query caused "No space left on device" error on my machine, where I had 200 GBs of empty space. I also would like to note that Citus planner is not even involved in below query execution as it's directly applied on shards.

```sql
SELECT count(1) AS count FROM (((((((test_schema.table1_dist_102069 dhc JOIN test_schema.table2_dist_102101 dscpctikcvced ON (((dscpctikcvced.m_hashkey)::text OPERATOR(pg_catalog.=) (dhc.m_hashkey)::text))) JOIN test_schema.table3_dist_102133 dscssi ON (((dscssi.m_hashkey)::text OPERATOR(pg_catalog.=) (dhc.m_hashkey)::text))) JOIN test_schema.table4_dist_102165 dlc2cp ON (((dlc2cp.m_hashkey)::text OPERATOR(pg_catalog.=) (dhc.m_hashkey)::text))) JOIN test_schema.table5_ref_102168 rhcp ON (((rhcp.m_hashkey)::text OPERATOR(pg_catalog.=) (dlc2cp.m_hashkey)::text))) JOIN test_schema.table6_ref_102169 rscpnas ON (((rscpnas.m_hashkey)::text OPERATOR(pg_catalog.=) (rhcp.m_hashkey)::text))) JOIN test_schema.table7_dist_102199 dlc2ca ON (((dlc2ca.m_hashkey)::text OPERATOR(pg_catalog.=) (dhc.m_hashkey)::text))) JOIN test_schema.table8_dist_102231 dhca ON (((dhca.m_hashkey)::text OPERATOR(pg_catalog.=) (dlc2ca.m_hashkey)::text))) WHERE true;
ERROR:  could not write to file "base/pgsql_tmp/pgsql_tmp9724.30": No space left on device
```

In summary, it doesn't seem like excessive memory / disk usage is due to Citus but the join generating  a huge amount of data.

**Comment by kot8891** [comment]
> First of all, the query plan cannot change by enabling / disabling repartition joins because the SELECT part of the query doesn't have a join that requires repartition join. But the whole insert .. select query requires "repartitioned insert .. select" (controlled via citus.enable_repartitioned_insert_select) because the query doesn't reference the distribution column of one of the source tables while inserting into distribution column of the target table.
> 
> So I'm closing the other issue while investigating this one further. (#7250)

sorry, didn't get it.
I have citus.enable_repartitioned_insert_select=on


So, I don't understand why running insert\ repartition-select cause out of RAM? Even if a query is not as optimal as it could be why it ends up with Postgresql instance fall instead of just some error of the query?

PS even if citus.enable_repartitioned_insert_select=off I have the same problem of leaking of RAM.

**Comment by onurctirtir** [comment]
> So, I don't understand why running insert\ repartition-select cause out of RAM? Even if a query is not as optimal as it could be why it ends up with Postgresql instance fall instead of just some error of the query?

I guess this depends on lots of configs in Postgres side because apparently this happens regardless of the fact that it's an insert .. select or plain select, and that, whether Citus is involved or not.

**Comment by kot8891** [comment]
> > So, I don't understand why running insert\ repartition-select cause out of RAM? Even if a query is not as optimal as it could be why it ends up with Postgresql instance fall instead of just some error of the query?
> 
> I guess this depends on lots of configs in Postgres side because apparently this happens regardless of the fact that it's an insert .. select or plain select, and that, whether Citus is involved or not.

I had this problem on postgresql 15, postgresql 16 with default config and postgresql 16 with custom config. 
Could you please to try to reproduce mem leak error https://github.com/citusdata/citus/issues/7250
